### PR TITLE
[mlir][memref] Fix emulate narrow types for strided memref offset

### PR DESCRIPTION
**Pull request by Groverkss**
This patch fixes strided memref offset calculation for emulating narrow types.

As a side effect, this patch also adds support for a 1-D subviews with static sizes, static offsets and strides of 1 for testing. Emulate narrow types pass was not tested for strided memrefs before this patch.

### CODE REVIEW

File: mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp [hunk 1]
```diff
@@ -209,6 +209,74 @@ struct ConvertMemRefLoad final : OpConversionPattern<memref::LoadOp> {
     return success();
   }
 };
+
+//===----------------------------------------------------------------------===//
+// ConvertMemRefSubview
+//===----------------------------------------------------------------------===//
+
+struct ConvertMemRefSubview final : OpConversionPattern<memref::SubViewOp> {
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp:217]: Please add some comments about limited support for subviews that are added here.

File: mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp [hunk 2]
```diff
@@ -209,6 +209,74 @@ struct ConvertMemRefLoad final : OpConversionPattern<memref::LoadOp> {
...
+    auto oldElementType = op.getSourceType().getElementType();
+    int srcBits = oldElementType.getIntOrFloatBitWidth();
+    int dstBits = convertedElementType.getIntOrFloatBitWidth();
+    if (dstBits % srcBits != 0) {
+      return rewriter.notifyMatchFailure(
+          op, "only dstBits % srcBits == 0 supported");
+    }
+
+    MemRefType newTy =
+        cast<MemRefType>(getTypeConverter()->convertType(op.getType()));
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp:235]: Use `dyn_cast`. Also you can probably just check this on the `op.getType()` directly.

File: mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp [hunk 3]
```diff
@@ -209,6 +209,74 @@ struct ConvertMemRefLoad final : OpConversionPattern<memref::LoadOp> {
...
+    auto size = op.getStaticSize(0);
+    auto offset = op.getStaticOffset(0);
+    // Only support static sizes and offsets.
+    if (size == ShapedType::kDynamic || offset == ShapedType::kDynamic) {
+      return rewriter.notifyMatchFailure(
+          op->getLoc(), "subview with dynamic size or offset is not supported");
+    }
+
+    int elementsPerByte = dstBits / srcBits;
+    if (size % elementsPerByte != 0 || offset % elementsPerByte != 0) {
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp:263]: I think you can drop the size constraint and make the size of the new `memref.subview` `ceilDiv(size, elementsPerByte)`.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just a few nits. Thanks!

File: mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp (hunk 2 above)
Comment (Groverkss) [reply mlir/lib/Dialect/MemRef/Transforms/EmulateNarrowType.cpp:235]: op.getType() returns MemRefType, but TypeConverter doesn't return the correct type sadly. I can make it dyn_cast.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**